### PR TITLE
Issue 442 improve cli push

### PR DESCRIPTION
**Pull request by ggalmazor**
Addresses #442:
- Avoid sending the blank form if Briefcase detects that it's already present in the destination server
- Add a new CLI flag to let users decide if they want to force sending the form anyway

While we implement the rest of #442, which involves changes on the GUI, pushing from it always sends the blank forms.

#### What has been done to verify that this works as intended?
- Added tests (unit and integration) for the HTTP interaction parts
- Manually tested the CLI and GUI to verify the intended behavior:
  - CLI outputs some message telling if it's sending the blank form or skipping it. Just run the CLI op with and without the `-fsb` flag
  - GUI should add the same lines as the CLI to the push details when used in combination with `-fsb`

#### Why is this the best possible solution? Were any other approaches considered?
Most of the changes in this PR are quite straightforward but I had to choose between using the existing HTTP interaction code or implementing it new.

I've chosen to implement a new `Http` port to avoid some of the problems we have already detected on the current code:
- It has side-effects that can be unpredictable effects, especially when combined with authentication
- It has a low level of abstraction, which can be verbose, and hard to follow and change
- Whatever code based on it is inherently harder to test

#### Are there any risks to merging this code? If so, what are they?
None.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Yep: https://github.com/opendatakit/docs/issues/703

### CODE REVIEW

File: test/java/org/opendatakit/briefcase/push/RemoteServerTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+package org.opendatakit.briefcase.push;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+import org.junit.Test;
+import org.opendatakit.briefcase.reused.http.FakeHttp;
+import org.opendatakit.briefcase.reused.http.Query;
+
+public class RemoteServerTest {
```
Comment (yanokwa) [inline test/java/org/opendatakit/briefcase/push/RemoteServerTest.java]: I don't think this test is valid and I think I led you astray with my curl command. You'll need to pass in a "X-OpenRosa-Version: 1.0" header..

`curl --header "X-OpenRosa-Version: 1.0" https://opendatakit.appspot.com/formList`

Your output will look like this:
```xml
<xforms xmlns="http://openrosa.org/xforms/xformsList">
<xform><formID>CascadingSelect</formID>
<name>Cascading Select Form</name>
<majorMinorVersion>2012072302</majorMinorVersion>
<version>2012072302</version>
<hash>md5:29a492009d35c1515886ddd307521819</hash>
<downloadUrl>https://opendatakit.appspot.com/formXml?formId=CascadingSelect</downloadUrl>
</xform>
<xform><formID>widgets</formID>
<name>Widgets</name>
<majorMinorVersion></majorMinorVersion>
<version></version>
<hash>md5:0eacc6333449e66826326eb5fcc75749</hash>
...
```

File: src/org/opendatakit/briefcase/push/RemoteServer.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * the License.
+ */
+
+package org.opendatakit.briefcase.push;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.opendatakit.briefcase.reused.http.Credentials;
```
Comment (yanokwa) [inline src/org/opendatakit/briefcase/push/RemoteServer.java]: What is the reason for the `reused` namespace?

File: test/java/org/opendatakit/briefcase/push/RemoteServerTest.java (hunk 1 above)
Comment (ggalmazor) [reply test/java/org/opendatakit/briefcase/push/RemoteServerTest.java]: OK, I've added the header and changed the test.

File: test/java/org/opendatakit/briefcase/push/RemoteServerTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            "<name>Some Form</name>\n" +
+            "<majorMinorVersion>2012072302</majorMinorVersion>\n" +
+            "<version>2012072302</version>\n" +
+            "<hash>md5:29a492009d35c1515886ddd307521819</hash>\n" +
+            "<downloadUrl>https://some.server.com/formXml?formId=some-form</downloadUrl>\n" +
+            "</xform>\n" +
+            "</xforms>"
+    );
+    RemoteServer server = RemoteServer.authenticated("https://some.server.com", "administrator", "aggregate");
+    assertThat(http.execute(server.containsFormQuery("some-form")), is(true));
```
Comment (dcbriccetti) [inline test/java/org/opendatakit/briefcase/push/RemoteServerTest.java]: You can ignore this, @ggalmazor, because I know I can’t persuade you, but this 'assert that x is true' instead of 'assertTrue x' bothers me even more than `if (boolean == true)`. OK, carry on.

Comment (ggalmazor) [reply test/java/org/opendatakit/briefcase/push/RemoteServerTest.java]: ¯\\\_(ツ)_/¯

File: src/org/opendatakit/briefcase/util/ServerUploader.java [hunk 4]
```diff
@@ -204,7 +212,14 @@ public boolean uploadFormAndSubmissionFiles(List<FormStatus> formsToTransfer) {
...
       boolean outcome;
-      outcome = uploadForm(formToTransfer, briefcaseFormDefFile, briefcaseFormMediaDir);
+      boolean existsAlready = http.execute(server.containsFormQuery(formToTransfer.getFormDefinition().getFormId()));
+      if (forceSendBlank || !existsAlready)
+        outcome = uploadForm(formToTransfer, briefcaseFormDefFile, briefcaseFormMediaDir);
+      else {
+        formToTransfer.setStatusString("Skipping form upload to remote server because it already exists", true);
+        EventBus.publish(new FormStatusEvent(formToTransfer));
+        outcome = true;
+      }
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/util/ServerUploader.java]: Below this: Bitwise **and** with booleans‽  Also:
screen shot 2018-05-15 at 12 10 13 pm

File: src/org/opendatakit/briefcase/push/RemoteServer.java (hunk 2 above)
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/push/RemoteServer.java]: Yes, the nuance being that you don't start off by creating stuff on `reused`. You move stuff there from a business layer when you need to use it from another business layer, thus it becomes _reused_

Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/push/RemoteServer.java]: Is this part of some well-known software architecture scheme?

File: src/org/opendatakit/briefcase/util/ServerUploader.java [hunk 5]
```diff
@@ -204,7 +212,14 @@ public boolean uploadFormAndSubmissionFiles(List<FormStatus> formsToTransfer) {
       File briefcaseFormMediaDir = FileSystemUtils.getMediaDirectoryIfExists(briefcaseLfd.getFormDirectory());
 
       boolean outcome;
-      outcome = uploadForm(formToTransfer, briefcaseFormDefFile, briefcaseFormMediaDir);
+      boolean existsAlready = http.execute(server.containsFormQuery(formToTransfer.getFormDefinition().getFormId()));
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/util/ServerUploader.java]: If you inline `existsAlready` or make it a function (lazy val in Scala), you can avoid the RPC when `forceSendBlank` is true, because the `if` will short circuit.

Comment (ggalmazor) [reply src/org/opendatakit/briefcase/util/ServerUploader.java]: It didn't occur to me making it lazy. Thanks for the suggestion!

**Comment by ggalmazor** [comment]
> Is this meant to come out twice?

You're getting the stdout CLI feedback message and the log lines, which are being redirected to stdout by logback. Changing the logback config will produce a different console output.

File: src/org/opendatakit/briefcase/reused/http/Credentials.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    Credentials that = (Credentials) o;
+    return Objects.equals(username, that.username) &&
+        Objects.equals(password, that.password);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(username, password);
```
Comment (dcbriccetti) [inline src/org/opendatakit/briefcase/reused/http/Credentials.java]: I see you’re using `Objects.hash`. Would `Objects.deepEquals` be useful, too?

Comment (ggalmazor) [reply src/org/opendatakit/briefcase/reused/http/Credentials.java]: I don't know about that. I override `.equals` and `.hash` to make Value Objects. Any member of a Value Object must be a primitive or another Value Object for this to work, therefore, `.equals` would and should be enough.

**Comment by kkrawczyk123** [comment]
### Tested with success!

Verified on Ubuntu, MacOS and Windows.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
